File: Python2/unique_morse_code_words.py
class Solution(object):
    def uniqueMorseRepresentations(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        def turn_into_morse(string):
            temp = []
            # O(len(string))
            # for ch in string:
            #     temp.append(morse[indexOf(ch)])  # amortized O(1)
            temp = [morse[indexOf(ch)] for ch in string] # use list comprehension will be better than normal for loop
            return ''.join(temp)  # O(len(output))

        def indexOf(char):  # O(1)
            return ord(char) - 97

        morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]

        myset = set()
        # counter = 0

        for w in words:
            morse_str = turn_into_morse(w)
            # if morse_str not in myset:  # O(1) in average, O(n) in worst
            #     myset.add(morse_str)  # set is implemented in hash table. O(1) if no collision otherwise O(n)
            #     counter += 1

            # add operation of set will keep the elements of set being unique
            myset.add(morse_str)

        # return counter
        return len(myset)


if __name__ == '__main__':
    words = ["gin", "zen", "gig", "msg"]
    sol = Solution()
    print sol.uniqueMorseRepresentations(words)

File: C++/unique_morse_code_words.cpp
#include <iostream>
#include <set>
#include "./unique_morse_code_words.h"
#include "./utility.h"

using namespace std;

string Solution::turn_into_morse(string str) {
    const char* morse[26] = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
    string output = "";
    // O(str.length()*O(?))
    for (int i = 0; i < str.length(); i++) {  // O(str.length())
        int idx = static_cast<int>(str[i])-97;  // O(1)
        output += morse[idx];  //O(?)  There are no standard complexity guarantees, typical implementations behave similar to std::vector::insert.
    }
    return output;
}

int Solution::uniqueMorseRepresentations(vector<string>& words) {
    set<string> myset;
    // O(words.size()*(str.length()*O(?) + logn))
    for (vector<string>::const_iterator iter = words.cbegin(); iter != words.cend(); ++iter) {  // O(words.size())
        string morse_str = turn_into_morse(*iter);  // O(str.length()*O(?))
        cout << morse_str << "\n";
        myset.insert(morse_str);  // O(logn)
    }

    return myset.size();
}

int main() {
    vector<string> words;
    words.push_back("gin");
    words.push_back("zen");
    words.push_back("gig");
    words.push_back("msg");

    Solution* sol = new Solution();
    int output = sol->uniqueMorseRepresentations(words);
    delete sol;

    cout << output << "\n";

    return 0;
}